### ISSUE TITLE
Runtime for bigger inputs

### ISSUE THREAD

**Issue by karabelaselias** [RobinMagnet/ScalableDenseMaps#1]
Hi there,

I was curious to use your code for some functional maps computations on bigger surface meshes. However I made a very simple test to check the timings of the dense maps calculations:

`import torch
from densemaps.torch import maps

emb1 = torch.randn((1,120000,128)).cuda().requires_grad_()
emb2 = torch.randn((1,120000,128)).cuda().requires_grad_()
emb1 = torch.nn.functional.normalize(
                emb1, p=2, dim=-1
            )
emb2 = torch.nn.functional.normalize(
                emb2, p=2, dim=-1
            )

map1 = maps.KernelDistMap(emb1, emb2, dist_type='sqdist', blur=0.1)

f = torch.randn((1,120000,384)).cuda().contiguous()

res = map1.pull_back(f)
loss = torch.mean(torch.square(res))
loss.backward(retain_graph=True)
`

If I run this forward and backward take combined around 35s. This is not really feasible in a case where I would embedd the dense map into another deepFM framework. Any ideas how this could be accelerated?

**Comment by RobinMagnet** [comment]
Hi,

I see the issue. This comes from KeOps being only fast for low memory operations. In particular, high dimensional distances are pretty slow to compute (with dimension 384 or 128). The number of points has little influence.

Since the multiplication `map1 @ f` is applied independently to each element of the last dimension of f, one could switch the last dimension to the batch dimension, which would definitely help.

If the issue is the dimension 128 of the embedding, then I believe the only solution is to do a manual loop sadly.

I can't do much tests right now but will check next week. Let me know if you find any fix.

**Comment by karabelaselias** [comment]
Hi Robin, thanks for taking a look, I was trying to use your maps in a context of rather dense surface meshes from medical applications. There was some work done in https://github.com/NafieAmrani/FUSS, however this approach fails rather spectaculary for denser meshes, which are essential for me to resolve all the intricate curvature of my domains. 

I tried their repo and tried to replace the dense similarity matrix with your code but this as I said doesn't really work, or more precisecly it does work but every iteration takes more than 40s which makes an ETA of 25 days for 15 epochs :D. 

So I'll take a look if I could work with independent dimensions and apply this in chunks or something. Otherwise I was toying around with approximating the similarity matrix using a topk approach and returning a sparse tensor, this of course is not exact but it can be done fast.

Cheers
Elias

**Comment by RobinMagnet** [comment]
Hi Elias,

I see. Since the issue if most probably related to dimensionality of the features, you can also try at some point to reduce the dimension before using dense maps.
The [benchmarks section of pykeops](https://www.kernel-operations.io/keops/_auto_benchmarks/index.html) might give you rough estimates on the ideal dimension. I believe that above dimension 100, naive pytorch can be faster (if it fits in memory).

I will try to post here an adapted version of the dense map, which will hopefully solve your issue. If not, I guess you will probably want to check with the keops developers.
